File: iosApp/src/Preview/Factories/EventFactory.swift
import Foundation
import Shared

#if DEBUG

enum EventFactory {

    private static let iso8601 = ISO8601DateFormatter()
    private static var now: String { iso8601.string(from: Date()) }

    private static func dateString(daysFromNow days: Int, hour: Int = 10) -> String {
        var components = DateComponents()
        components.day = days
        components.hour = hour - Calendar.current.component(.hour, from: Date())
        let date = Calendar.current.date(byAdding: components, to: Date()) ?? Date()
        return iso8601.string(from: date)
    }

    private static func slot(
        id: String,
        start: String?,
        end: String?,
        timezone: String = "Europe/Paris",
        timeOfDay: TimeOfDay = .specific
    ) -> TimeSlot_ {
        TimeSlot_(id: id, start: start, end: end, timezone: timezone, timeOfDay: timeOfDay)
    }

    static var empty: Event_ {
        make(
            id: "event-empty-001",
            title: "",
            description: "",
            status: .draft,
            eventType: Shared.EventType.other
        )
    }

    static var complete: Event_ {
        make(
            id: "event-complete-002",
            title: "Weekend Hiking Trip",
            description: "Join us for a scenic hike through the Calanques National Park followed by a picnic.",
            participants: UserFactory.group(count: 5).map(\.id),
            proposedSlots: sampleTimeSlots,
            deadline: dateString(daysFromNow: 3),
            status: .confirmed,
            finalDate: dateString(daysFromNow: 14, hour: 9),
            createdAt: dateString(daysFromNow: -7),
            eventType: Shared.EventType.outdoorActivity,
            minParticipants: 3,
            maxParticipants: 20,
            expectedParticipants: 8,
            heroImageUrl: "https://images.unsplash.com/photo-1551632811-561732d1e306?w=800"
        )
    }

    static var past: Event_ {
        make(
            id: "event-past-003",
            title: "Team Offsite 2025",
            description: "Annual team building weekend in the Alps.",
            participants: UserFactory.group(count: 4).map(\.id),
            deadline: dateString(daysFromNow: -30),
            status: .finalized,
            finalDate: dateString(daysFromNow: -14, hour: 10),
            createdAt: dateString(daysFromNow: -60),
            updatedAt: dateString(daysFromNow: -14),
            eventType: Shared.EventType.teamBuilding,
            minParticipants: 4,
            expectedParticipants: 6
        )
    }

    static var polling: Event_ {
        make(
            id: "event-polling-004",
            title: "Birthday Dinner",
            description: "Celebrating Emma's birthday! Vote for the best date.",
            participants: UserFactory.group(count: 3).map(\.id),
            proposedSlots: [
                slot(id: "slot-poll-1", start: dateString(daysFromNow: 7, hour: 19), end: dateString(daysFromNow: 7, hour: 22), timeOfDay: .evening),
                slot(id: "slot-poll-2", start: dateString(daysFromNow: 8, hour: 12), end: dateString(daysFromNow: 8, hour: 14), timeOfDay: .afternoon),
                slot(id: "slot-poll-3", start: nil, end: nil, timeOfDay: .morning)
            ],
            deadline: dateString(daysFromNow: 5),
            status: .polling,
            createdAt: dateString(daysFromNow: -2),
            eventType: Shared.EventType.birthday,
            minParticipants: 5,
            maxParticipants: 15,
            expectedParticipants: 10
        )
    }

    static var withManyParticipants: Event_ {
        make(
            id: "event-crowd-005",
            title: "Company Summer Party",
            description: "End-of-summer celebration with BBQ, games, and live music at Parc Borély.",
            participants: (0..<18).map { "user-crowd-\($0)" },
            deadline: dateString(daysFromNow: 10),
            status: .confirmed,
            finalDate: dateString(daysFromNow: 21, hour: 16),
            createdAt: dateString(daysFromNow: -14),
            eventType: Shared.EventType.party,
            minParticipants: 10,
            maxParticipants: 50,
            expectedParticipants: 25,
            heroImageUrl: "https://images.unsplash.com/photo-1533174072545-7a4b6ad7a6c3?w=800"
        )
    }

    static func make(
        id: String = "event-\(UUID().uuidString.prefix(8))",
        title: String = "Test Event",
        description: String = "A test event for preview purposes.",
        organizerId: String = UserFactory.organizer.id,
        participants: [String] = [],
        proposedSlots: [TimeSlot_] = [],
        deadline: String? = nil,
        status: EventStatus = .draft,
        finalDate: String? = nil,
        createdAt: String? = nil,
        updatedAt: String? = nil,
        eventType: Shared.EventType = Shared.EventType.other,
        eventTypeCustom: String? = nil,
        minParticipants: Int32? = nil,
        maxParticipants: Int32? = nil,
        expectedParticipants: Int32? = nil,
        heroImageUrl: String? = nil
    ) -> Event_ {
        let nowString = now
        return Event_(
            id: id,
            title: title,
            description: description,
            organizerId: organizerId,
            participants: participants,
            proposedSlots: proposedSlots,
            deadline: deadline ?? dateString(daysFromNow: 7),
            status: status,
            finalDate: finalDate,
            createdAt: createdAt ?? nowString,
            updatedAt: updatedAt ?? nowString,
            eventType: eventType,
            eventTypeCustom: eventTypeCustom,
            minParticipants: minParticipants.map { KotlinInt(value: $0) },
            maxParticipants: maxParticipants.map { KotlinInt(value: $0) },
            expectedParticipants: expectedParticipants.map { KotlinInt(value: $0) },
            heroImageUrl: heroImageUrl
        )
    }

    private static var sampleTimeSlots: [TimeSlot_] {
        [
            slot(id: "slot-sample-1", start: dateString(daysFromNow: 14, hour: 9), end: dateString(daysFromNow: 14, hour: 17), timeOfDay: .specific),
            slot(id: "slot-sample-2", start: nil, end: nil, timeOfDay: .morning)
        ]
    }
}

#endif

File: composeApp/src/androidMain/kotlin/com/guyghost/wakeve/preview/factories/EventFactory.kt
package com.guyghost.wakeve.preview.factories

import com.guyghost.wakeve.models.Event
import com.guyghost.wakeve.models.EventStatus
import com.guyghost.wakeve.models.EventType
import com.guyghost.wakeve.models.TimeOfDay
import com.guyghost.wakeve.models.TimeSlot

/**
 * Factory for creating preview/test Event instances.
 *
 * All properties use `get()` to return fresh instances on each access.
 * Dates use ISO 8601 string literals for simplicity.
 */
object EventFactory {

    // ---- Reusable time slots ------------------------------------------------

    private val futureSlotA
        get() = TimeSlot(
            id = "slot-001",
            start = "2026-04-12T10:00:00Z",
            end = "2026-04-12T18:00:00Z",
            timezone = "Europe/Paris",
            timeOfDay = TimeOfDay.SPECIFIC
        )

    private val futureSlotB
        get() = TimeSlot(
            id = "slot-002",
            start = "2026-04-19T14:00:00Z",
            end = "2026-04-19T22:00:00Z",
            timezone = "Europe/Paris",
            timeOfDay = TimeOfDay.AFTERNOON
        )

    private val futureSlotC
        get() = TimeSlot(
            id = "slot-003",
            start = "2026-04-26T09:00:00Z",
            end = "2026-04-26T17:00:00Z",
            timezone = "Europe/Paris",
            timeOfDay = TimeOfDay.MORNING
        )

    private val pastSlotA
        get() = TimeSlot(
            id = "slot-past-001",
            start = "2025-12-20T18:00:00Z",
            end = "2025-12-20T23:00:00Z",
            timezone = "Europe/Paris",
            timeOfDay = TimeOfDay.EVENING
        )

    private val pastSlotB
        get() = TimeSlot(
            id = "slot-past-002",
            start = "2025-12-21T10:00:00Z",
            end = "2025-12-21T16:00:00Z",
            timezone = "Europe/Paris",
            timeOfDay = TimeOfDay.SPECIFIC
        )

    // ---- Factory properties -------------------------------------------------

    /** An empty draft event with no participants or slots. */
    val empty
        get() = Event(
            id = "event-draft-001",
            title = "",
            description = "",
            organizerId = "user-organizer-001",
            participants = emptyList(),
            proposedSlots = emptyList(),
            deadline = "2026-04-01T23:59:59Z",
            status = EventStatus.DRAFT,
            createdAt = "2026-03-01T08:00:00Z",
            updatedAt = "2026-03-01T08:00:00Z",
            eventType = EventType.OTHER
        )

    /** A fully confirmed event with 5 participants and 2 time slots. */
    val complete
        get() = Event(
            id = "event-confirmed-002",
            title = "Weekend Hiking Trip",
            description = "A two-day hike in the Vosges mountains with overnight camping.",
            organizerId = "user-organizer-001",
            participants = listOf(
                "user-participant-002",
                "user-email-003",
                "user-group-004",
                "user-group-005",
                "user-group-006"
            ),
            proposedSlots = listOf(futureSlotA, futureSlotB),
            deadline = "2026-04-05T23:59:59Z",
            status = EventStatus.CONFIRMED,
            finalDate = "2026-04-12T10:00:00Z",
            createdAt = "2026-03-01T10:00:00Z",
            updatedAt = "2026-03-10T14:30:00Z",
            eventType = EventType.OUTDOOR_ACTIVITY,
            minParticipants = 3,
            maxParticipants = 10,
            expectedParticipants = 6,
            heroImageUrl = null
        )

    /** An event in polling phase with 8 participants and 3 proposed slots. */
    val polling
        get() = Event(
            id = "event-polling-003",
            title = "Team Offsite Q2",
            description = "Quarterly team-building day: activities, lunch, and retrospective.",
            organizerId = "user-organizer-001",
            participants = (1..8).map { "user-polling-${it.toString().padStart(3, '0')}" },
            proposedSlots = listOf(futureSlotA, futureSlotB, futureSlotC),
            deadline = "2026-04-08T23:59:59Z",
            status = EventStatus.POLLING,
            createdAt = "2026-03-02T09:00:00Z",
            updatedAt = "2026-03-03T11:00:00Z",
            eventType = EventType.TEAM_BUILDING,
            expectedParticipants = 8
        )

    /** A finalized event with dates in the past. */
    val past
        get() = Event(
            id = "event-past-004",
            title = "Holiday Dinner 2025",
            description = "End-of-year team dinner at Le Petit Bistrot.",
            organizerId = "user-organizer-001",
            participants = listOf(
                "user-participant-002",
                "user-email-003",
                "user-group-004"
            ),
            proposedSlots = listOf(pastSlotA, pastSlotB),
            deadline = "2025-12-15T23:59:59Z",
            status = EventStatus.FINALIZED,
            finalDate = "2025-12-20T18:00:00Z",
            createdAt = "2025-11-20T12:00:00Z",
            updatedAt = "2025-12-20T23:00:00Z",
            eventType = EventType.PARTY,
            expectedParticipants = 4
        )

    /**
     * Returns a mixed list of events covering all key statuses.
     */
    fun mixedList(): List<Event> = listOf(empty, polling, complete, past)
}
